Breakdown jobs and rbac templates in Helm Chart

adityajalkhare opened the pull request:
# Description

The job templates and rbac resources templates are broken down into multiple files instead of a single file.

This PR is with respect to: https://github.com/solo-io/gloo/issues/8616

## Code changes

**5-resource-cleanup-job**

| Template                                       | Description                         |
|------------------------------------------------|-------------------------------------|
| 5-resource-cleanup-job.yaml                    | Has only `Job` spec                 |
| 5-resource-cleanup-job-service-account.yaml    | Has only `Service Account` spec     | 
| 5-resource-cleanup-job-role.yaml               | Has only `Role` spec                | 
| 5-resource-cleanup-job-rolebinding.yaml        | Has only `RoleBinding` spec         |
| 5-resource-cleanup-job-clusterrole.yaml        | Has only `ClusterRole` spec         |
| 5-resource-cleanup-job-clusterrolebinding.yaml | Has only `ClusterRoleBinding` spec  |

**5-resource-migration-job**

| Template                                         | Description                         |
|--------------------------------------------------|-------------------------------------|
| 5-resource-migration-job.yaml                    | Has only `Job` spec                 |
| 5-resource-migration-job-service-account.yaml    | Has only `Service Account` spec     | 
| 5-resource-migration-job-role.yaml               | Has only `Role` spec                | 
| 5-resource-migration-job-rolebinding.yaml        | Has only `RoleBinding` spec         |
| 5-resource-migration-job-clusterrole.yaml        | Has only `ClusterRole` spec         |
| 5-resource-migration-job-clusterrolebinding.yaml | Has only `ClusterRoleBinding` spec  |

**5-resource-rollout-job**

| Template                                       | Description                         |
|------------------------------------------------|-------------------------------------|
| 5-resource-rollout-job.yaml                    | Has only `Job` spec                 |
| 5-resource-rollout-job-service-account.yaml    | Has only `Service Account` spec     | 
| 5-resource-rollout-job-role.yaml               | Has only `Role` spec                | 
| 5-resource-rollout-job-rolebinding.yaml        | Has only `RoleBinding` spec         |
| 5-resource-rollout-job-clusterrole.yaml        | Has only `ClusterRole` spec         |
| 5-resource-rollout-job-clusterrolebinding.yaml | Has only `ClusterRoleBinding` spec  |

**6.5-gateway-certgen-job**

| Template                                                | Description                                      |
|---------------------------------------------------------|--------------------------------------------------|
|6.5-gateway-certgen-job.yaml                             | Has only `Job` spec                              |
|6.5-gateway-certgen-job-service-account.yaml    | Has only `Service Account` spec     | 
|6.5-gateway-certgen-job-secret-create-clusterrole        | Has only `secret-create` `(Cluster)Role` spec        |
|6.5-gateway-certgen-job-secret-create-clusterrolebinding | Has only `secret-create` `(Cluster)RoleBinding` spec |
|6.5-gateway-certgen-job-vwc-update-clusterrole           | Has only vwc-update `ClusterRole` spec           |
|6.5-gateway-certgen-job-vwc-update-clusterrolebinding    | Has only vwc-update `ClusterRoleBinding` spec    |


# Context

## Interesting decisions
 
We chose to do things this way because:
1. Breaking out each resource definition in its own template file is the [recommended way](https://helm.sh/docs/chart_best_practices/templates/#structure-of-templates).
2. For finer control over the resources.
3. In case someone wants to fork and create cluster-level resources and namespace-scoped resources separately, this change will help control that in a better way.

## Testing steps

None

## Notes for reviewers

None


# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas - N/A
- [x] I have made corresponding changes to the documentation - N/A
- [ ] I have added tests that prove my fix is effective or that my feature works - N/A

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/8616

jenshu commented:
@adityajalkhare thanks for the contribution. Could you please add a changelog file? This would go into the next 1.16.x [release](https://github.com/solo-io/gloo/releases), so v1.16.0-beta5.

The changes seem reasonable; I verified that there were no changes to the content, and that they were simply moved into new files. Just a note that in some of the other gloo helm templates, we further break them down into serviceaccount, clusterrole, clusterrolebinding, etc files, although I don't necessarily think you need to make that change here.

adityajalkhare commented:
@jenshu 
> we further break them down into serviceaccount, clusterrole, clusterrolebinding, etc files

This is what I was going to ask. I am planning on breaking them down further for granularity. It'll be much better.

adityajalkhare commented:
@jenshu Kindly review and approve :)

jenshu commented:
/test

adityajalkhare commented:
@jenshu one of the tests (`build-bot`) is failing. Let me know what can be done to fix that.

adityajalkhare commented:
/test

adityajalkhare commented:
@jenshu Please review

jenshu commented:
@adityajalkhare builds are currently failing from forks, we are looking into it. we will need to fix that before your build will pass

adityajalkhare commented:
@jenshu, thanks for the inputs. I have updated the changelog into `v1.16.0-beta6`.

Also added reasoning behind this change in both PR Description and [issue description](https://github.com/solo-io/gloo/issues/8616)

jenshu commented:
/test

bewebi commented on install/helm/gloo/templates/6.5-gateway-certgen-job-clusterrole.yaml:
```diff
@@ -0,0 +1,38 @@
+  resources: ["validatingwebhookconfigurations"]
+  verbs: ["get", "update"]
+---
+kind: {{ include "gloo.roleKind" . }}
```
Should we split this resource off into it's own file as well?

Whether or not it's a Role or a ClusterRole depends on the [`global.glooRbac.namespaced` value](https://github.com/solo-io/gloo/blob/96be00455720b826e7f94a934482b7f3d901c32a/install/helm/gloo/templates/_helpers.tpl#L4) so naming for that file might need to be ambiguous, but I think it would follow on the same principle as the rest of the PR that the (Cluster)Role should be separated

adityajalkhare replied:
@bewebi I was confused whether to do that or not.
Thanks for the input! These changes are now done!

bewebi commented on install/helm/gloo/templates/6.5-gateway-certgen-job-clusterrolebinding.yaml:
```diff
@@ -0,0 +1,45 @@
+  name: gloo-gateway-vwc-update{{ include "gloo.rbacNameSuffix" . }}
+  apiGroup: rbac.authorization.k8s.io
+---
+kind: {{ include "gloo.roleKind" . }}Binding
```
See comment above

bewebi reviewed (commented):
It seems like we can further factor out a couple more resource files

adityajalkhare commented:
/test

adityajalkhare commented:
@jenshu @bewebi please review

bewebi reviewed (approved):
With the secret (Cluster)Role files I'm not sure we're going to be able to do better naming-wise than using the kind we'll generate by default, which is ClusterRole
